agent/proxycfg/api_gateway.go

```diff
@@ -0,0 +1,492 @@
+					// to force that protocol on the Envoy listener.
+					Config: map[string]interface{}{
+						"protocol": "tcp",
+					},
+				}
+
+				listenerKey := APIGatewayListenerKey{Protocol: string(listener.Protocol), Port: listener.Port}
+				upstreams[listenerKey] = append(upstreams[listenerKey], upstream)
+			}
+
+			watchOpts := discoveryChainWatchOpts{
+				id:         upstreamID,
+				name:       service.Name,
+				namespace:  service.NamespaceOrDefault(),
+				partition:  service.PartitionOrDefault(),
+				datacenter: h.stateConfig.source.Datacenter,
+			}
+
+			handler := &handlerUpstreams{handlerState: h.handlerState}
+			if err := handler.watchDiscoveryChain(ctx, snap, watchOpts); err != nil {
+				return fmt.Errorf("failed to watch discovery chain for %s: %w", upstreamID, err)
+			}
+		}
+	default:
+		return fmt.Errorf("invalid type for config entry: %T", resp.Entry)
+	}
+
+	snap.APIGateway.Upstreams = upstreams
+	snap.APIGateway.UpstreamsSet = seenUpstreamIDs
+	//snap.APIGateway.Hosts = TODO
+	snap.APIGateway.HostsSet = true
+
+	// Stop watching any upstreams and discovery chains that have become irrelevant
+	for upstreamID, cancelDiscoChain := range snap.APIGateway.WatchedDiscoveryChains {
+		if _, ok := seenUpstreamIDs[upstreamID]; ok {
+			continue
+		}
+
+		for targetID, cancelUpstream := range snap.APIGateway.WatchedUpstreams[upstreamID] {
+			cancelUpstream()
+			delete(snap.APIGateway.WatchedUpstreams[upstreamID], targetID)
+			delete(snap.APIGateway.WatchedUpstreamEndpoints[upstreamID], targetID)
+
+			if targetUID := NewUpstreamIDFromTargetID(targetID); targetUID.Peer != "" {
+				snap.APIGateway.PeerUpstreamEndpoints.CancelWatch(targetUID)
+				snap.APIGateway.UpstreamPeerTrustBundles.CancelWatch(targetUID.Peer)
+			}
+		}
+
+		cancelDiscoChain()
+		delete(snap.APIGateway.WatchedDiscoveryChains, upstreamID)
+	}
+
+	return nil
+}
+
+// referenceIsForListener returns whether the provided structs.ResourceReference
+// targets the provided structs.APIGatewayListener. For this to be true, the kind
+// and name must match the structs.APIGatewayConfigEntry containing the listener,
+// and the reference must specify either no section name or the name of the listener
+// as the section name.
+func (h *handlerAPIGateway) referenceIsForListener(ref structs.ResourceReference, listener structs.APIGatewayListener, snap *ConfigSnapshot) bool {
+	if ref.Kind != snap.APIGateway.GatewayConfig.Kind {
+		return false
+	}
+	if ref.Name != snap.APIGateway.GatewayConfig.Name {
+		return false
+	}
+	return ref.SectionName == "" || ref.SectionName == listener.Name
+}
+
+func (h *handlerAPIGateway) watchIngressLeafCert(ctx context.Context, snap *ConfigSnapshot) error {
+	// Note that we DON'T test for TLS.enabled because we need a leaf cert for the
+	// gateway even without TLS to use as a client cert.
+	if !snap.IngressGateway.GatewayConfigLoaded || !snap.IngressGateway.HostsSet {
```

Comment:
I'm assuming this should be `APIGateway` v. `IngressGateway`? Also, seeing as serving over TLS requires specifying an inline certificate at the moment, I think we should probably drop this entire handler -- it seems to be for pulling the terminating TLS cert signed by the connect CA.